Adds photo upload description info per bug 38282 and bug 38286

awjrichards-zz opened the pull request:
- Modifies python script to capture full campaign data rather than limited subset
- Python script now stores campaign data in a dict before dumping to json
- Updates campaigns-data.js to latest full campaign data
- Campaign data structure now made up of keyed objects, campaign name is key
- Campaign sorting by alpha now happens in the template
- upload.js now required in app.js
- updated formatUploadDescription() to specs in bug 38282
- Adds specific categories per bug 38286, with the exception of UA, which
  will be dealt with later.

Note that this is a replacement for previous commmits that I've abandoned
due to some serious git fu fail on my part.

awjrichards-zz commented:
See sample upload at http://test.wikipedia.org/w/index.php?title=File:Marist_College_Historic_District_1342124590972.jpeg&action=edit

jdlrobson commented on assets/www/index.html:
```diff
@@ -16,7 +16,7 @@
   <body class='hidden'>
 	<script type="text/html" id="country-list-template">
-		<% _.each(countries, function( country ) { %>
+		<% _.each( _.sortBy( countries, function(country) { return country.desc }), function( country ) { %>
```
whitespace needed after .desc }

yuvipanda replied:
I see whitespace after .desc :P

jdlrobson replied:
After the curly brace..

jdlrobson commented on assets/www/js/app.js:
```diff
@@ -223,10 +223,14 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'preference
 		var comment = 'Uploaded via WLM Mobile App';
-		var text = 'Testing WLM';
 		var uploadConfirmTemplate = templates.getTemplate('upload-confirm-template');
+
```
unrelated whitespace

jdlrobson commented on assets/www/js/app.js:
```diff
@@ -223,10 +223,14 @@ require( [ 'jquery', 'l10n', 'geo', 'api', 'templates', 'monuments', 'preference
 		var fileName = curMonument.generateFilename();
 		console.log("Filename is " + fileName);
+
+		var text = formatUploadDescription( curMonument, CAMPAIGNS[curMonument.country].config, api.userName );
```
whitespace around curMonument

jdlrobson commented on assets/www/js/upload.js:
```diff
@@ -46,39 +46,49 @@ function dateYMD() {
+		autoWikiText = campaignConfig.autoWikiText;
+
 	var desc = '';
 	desc += '=={{int:filedesc}}==\n';
```
can we do this in a more generic way? e.g. make sure of templates - it would be good to allow easy changes to this

jdlrobson commented on assets/www/js/upload.js:
```diff
@@ -46,39 +46,49 @@ function dateYMD() {
+			'Uploaded with Android WLM App'
+		],
+		cats = campaignConfig.defaultCategories
+			.concat( campaignConfig.autoCategories )
```
we usually trail the dot '.' elsewhere in the app e.g.

```
cats = campaignConfig.defaultCategories.
concat
```

yuvipanda replied:
Do we need to do that at all? I personally prefer not breaking them up.

jdlrobson commented on scripts/generate_campaigns.py:
```diff
@@ -14,20 +14,23 @@
-				'licenses': camp['config']['licensesOwnWork']
-				}
-		if campaign['name'] in namesdata:
+		camp['name'] = camp['name'].replace('wlm-','')
```
use more convenient dot notation here ( camp.name ) - only use bracketed quotes when  you need to (e.g. variable/reserved keyword/has a - or other keyword character)

awjrichards-zz replied:
dude, this is a python script, not js :)

jdlrobson replied:
Lolz. Fail.

jdlrobson commented on scripts/generate_campaigns.py:
```diff
@@ -14,20 +14,23 @@
 			campaign['desc'] = namesdata[campaign['name']]
-			# Only pickup campaings that have a description set
-			campaigns.append(campaign)
+			campaigns.append( campaign )
```
unrelated change

jdlrobson commented:
Could we separate this into 2 commits? One for changing the format of CAMPAIGN and one for the description - As Yuvi will testify I like smaller commit for when I need to debug :).

jdlrobson commented:
Works for me. On test wiki the templates do not get rendered - but I assume that is because test wiki is not setup.

awjrichards-zz committed:
Updates campaign info generation and scheme to use full campaign data from API
* Modifies python script to capture full campaign data rather than limited subset
* Python script now stores campaign data in a dict before dumping to json
* Updates campaigns-data.js to latest full campaign data
* Campaign data structure now made up of keyed objects, campaign name is key

awjrichards-zz committed:
Adds photo upload description info
* Campaign sorting by alpha now happens in the template
* upload.js now required in app.js
* updated formatUploadDescription() to specs in bug 38282
* Adds specific categories per bug 38286, with the exception of UA, which will be dealt with later.

awjrichards-zz commented:
I split the commit into two commits and addressed most of Jon's issues with the exception of using templates for the photo description text.

awjrichards-zz commented:
The main reason I haven't implemented the upload description as a template is because I haven't worked with js/templates much before, and certainly not in this kind of context. If you or someone could give me a little guidance, I'd be happy to implement it.

awjrichards-zz committed:
Adds android version and device as category to photoupload
* Uses android version and device subset of UA as add'l category

awjrichards-zz commented:
Ok - I did some reading on underscore and tried implementing a template instead of concat'ing 'desc'. It does not appear to be working, however. You can see the diff of my changes here:

https://gist.github.com/3102170

After rebuilding the app with these changes and trying to upload a photo, I can take a photo or select one from the device, but then I am taken back to the 'Add a photo' screen. logcat shows:
`D/CordovaLog( 8621): Error in success callback: Camera4 = ReferenceError: formatUploadDescription is not defined`
`D/CordovaLog( 8621): file:///android_asset/www/android/cordova-1.8.1.js: Line 254 : Error in success callback: Camera4 = ReferenceError: formatUploadDescription is not defined`

jdlrobson commented:
Updating gist with comments.  ReferenceError is probably due to no requires statement in upload.js - it seems that previously this file has not been used and predates the use of require js.

awjrichards-zz committed:
Adds template support for photo upload description
* Adds pohoto desc template to index.html
* Moves description format and date handling functions out of
upload.js

awjrichards-zz committed:
Makes template caching optional in getTemplate()
This is needed for templates like photo description upload which
should not get cached.

awjrichards-zz commented:
Got the template working properly. Proof is in the pudding: http://test.wikipedia.org/wiki/File:Antiga_f%C3%A0brica_de_pells_(Escaldes-Engordany)_1342487002478.jpeg

awjrichards-zz committed:
Removing upload.js as it is no longer needed.
